In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns

In [2]:
pt = pd.read_csv('ProjectTycho_Level2_v1.1.0.csv').head(100000)

In [3]:
pt['DrC'] = pt.iloc[:,6]

In [4]:
pt.drop('url', axis = 1, inplace = True)

In [5]:
pt['Year']=[d.split('-')[0] for d in pt.from_date]
pt['Month']=[d.split('-')[1] for d in pt.from_date]
pt['Day']=[d.split('-')[2] for d in pt.from_date]

In [6]:
pt.head()

epi_week country state           loc loc_type  \
0    188824      US    PA  PHILADELPHIA     CITY   
1    188824      US    PA  PHILADELPHIA     CITY   
2    188824      US    PA  PHILADELPHIA     CITY   
3    188826      US    PA  PHILADELPHIA     CITY   
4    188826      US    PA  PHILADELPHIA     CITY   

                         disease   event  number   from_date     to_date  \
0  TYPHOID FEVER [ENTERIC FEVER]  DEATHS      14  1888-06-10  1888-06-16   
1                  SCARLET FEVER  DEATHS       4  1888-06-10  1888-06-16   
2                     DIPHTHERIA  DEATHS       4  1888-06-10  1888-06-16   
3  TYPHOID FEVER [ENTERIC FEVER]  DEATHS      12  1888-06-24  1888-06-30   
4                  SCARLET FEVER  DEATHS       5  1888-06-24  1888-06-30   

      DrC  Year Month Day  
0  DEATHS  1888    06  10  
1  DEATHS  1888    06  10  
2  DEATHS  1888    06  10  
3  DEATHS  1888    06  24  
4  DEATHS  1888    06  24

In [7]:
pt['DrC'] = pd.get_dummies(pt['DrC'], drop_first=True)

In [8]:
pt.head(10)

epi_week country state           loc loc_type  \
0    188824      US    PA  PHILADELPHIA     CITY   
1    188824      US    PA  PHILADELPHIA     CITY   
2    188824      US    PA  PHILADELPHIA     CITY   
3    188826      US    PA  PHILADELPHIA     CITY   
4    188826      US    PA  PHILADELPHIA     CITY   
5    188826      US    PA  PHILADELPHIA     CITY   
6    188826      US    MD     BALTIMORE     CITY   
7    188826      US    MD     BALTIMORE     CITY   
8    188825      US    LA   NEW ORLEANS     CITY   
9    188825      US    LA   NEW ORLEANS     CITY   

                         disease   event  number   from_date     to_date  DrC  \
0  TYPHOID FEVER [ENTERIC FEVER]  DEATHS      14  1888-06-10  1888-06-16    1   
1                  SCARLET FEVER  DEATHS       4  1888-06-10  1888-06-16    1   
2                     DIPHTHERIA  DEATHS       4  1888-06-10  1888-06-16    1   
3  TYPHOID FEVER [ENTERIC FEVER]  DEATHS      12  1888-06-24  1888-06-30    1   
4                  SCARLET FEVER  DEATHS       5  1888-06-24  1888-06-30    1   
5                     DIPHTHERIA  DEATHS       7  1888-06-24  1888-06-30    1   
6  TYPHOID FEVER [ENTERIC FEVER]  DEATHS       4  1888-06-24  1888-06-30    1   
7                     DIPHTHERIA  DEATHS       1  1888-06-24  1888-06-30    1   
8  TYPHOID FEVER [ENTERIC FEVER]  DEATHS       1  1888-06-17  1888-06-23    1   
9                     DIPHTHERIA  DEATHS       4  1888-06-17  1888-06-23    1   

   Year Month Day  
0  1888    06  10  
1  1888    06  10  
2  1888    06  10  
3  1888    06  24  
4  1888    06  24  
5  1888    06  24  
6  1888    06  24  
7  1888    06  24  
8  1888    06  17  
9  1888    06  17

In [9]:
pt.number.unique()

array([  14,    4,   12,    5,    7,    1,    9,    2,    6,    3,   13,
         17,   19,   35,   20,   23,    8,   10,   11,   22,   27,   15,
         37,   40,   24,   16,   32,   25,   41,   18,   29,   21,   39,
         26,   33,   36,   30,   34,   52,   48,   51,   47,   43,   49,
         54,   50,   55,   44,   38,   28,   58,   46,   31,   59,   42,
         57,   71,   53,   61,   56,    0,   63,  104,  119,  102,  124,
        137,   64,  118,   78,  127,   60,  114,   45,   99,  107,  112,
        108,  103,   88,   66,   98,   85,   70,   72,   89,   82,   80,
         74,  116,   73,   77,   76,   87,   95,   62,  101,   97,   86,
        110,   90,   69,  120,   79,  129,  132,  105,  128,  121,   93,
        111,   92,   83,   67,   75,   84,  106,   94,   96,  149,  135,
        133,  123,   68,   65,   81,  100,   91,  122,  109,  115,  117,
        134,  140,  160,  158,  166,  148,  159,  151,  153,  168,  144,
        155,  190, 4127,  136,  138,  157,  145,  1

In [10]:
len(pt.Month.unique())

12

In [11]:
len(pt.state.unique())

43

In [12]:
len(pt.disease.unique())

14

In [13]:
len(pt.DrC.unique())

2

In [14]:
x = pt.iloc[:,[2,4,5,7,10,12,13]]

In [15]:
x.head()

state loc_type                        disease  number  DrC Month Day
0    PA     CITY  TYPHOID FEVER [ENTERIC FEVER]      14    1    06  10
1    PA     CITY                  SCARLET FEVER       4    1    06  10
2    PA     CITY                     DIPHTHERIA       4    1    06  10
3    PA     CITY  TYPHOID FEVER [ENTERIC FEVER]      12    1    06  24
4    PA     CITY                  SCARLET FEVER       5    1    06  24

In [16]:
y = pt.iloc[:,3]

In [17]:
y.head()

0    PHILADELPHIA
1    PHILADELPHIA
2    PHILADELPHIA
3    PHILADELPHIA
4    PHILADELPHIA
Name: loc, dtype: object

In [18]:
a = pd.get_dummies(x['state'], drop_first=True )
b = pd.get_dummies(x['loc_type'], drop_first=True )
c = pd.get_dummies(x['disease'], drop_first=True )

In [19]:
cmb = pd.concat([a,b], axis = 1)
cmb = pd.concat([cmb,c], axis = 1)
cmb = pd.concat([cmb,pt['DrC']], axis = 1)
cmb = pd.concat([cmb,pt['number']], axis = 1)
cmb = pd.concat([cmb,pt['Month']], axis = 1)
cmb = pd.concat([cmb,pt['Day']], axis = 1)

In [20]:
cmb.shape

(100000, 60)

In [21]:
import sklearn

In [22]:
from sklearn import preprocessing

In [23]:
from sklearn.linear_model import LogisticRegression

### Logistic Regression

In [24]:
lr = LogisticRegression()
from sklearn.cross_validation import train_test_split
x_train, x_test, y_train, y_test = train_test_split(cmb,y, test_size = 0.3)


/home/student/anaconda3/lib/python3.6/site-packages/sklearn/cross_validation.py:41: DeprecationWarning: This module was deprecated in version 0.18 in favor of the model_selection module into which all the refactored classes and functions are moved. Also note that the interface of the new CV iterators are different from that of this module. This module will be removed in 0.20.
  "This module will be removed in 0.20.", DeprecationWarning)


In [25]:
lr.fit(x_train,y_train)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l2', random_state=None, solver='liblinear', tol=0.0001,
          verbose=0, warm_start=False)

In [27]:
y_pred = lr.predict(x_test)

In [28]:
from sklearn.metrics import confusion_matrix
confusion_matrix(y_test, y_pred)

array([[ 0,  0,  0, ...,  0,  0,  0],
       [ 0,  0,  0, ...,  0,  0,  0],
       [ 0,  0,  0, ...,  0,  0,  0],
       ..., 
       [ 0,  0,  0, ..., 10,  0,  0],
       [ 0,  0,  0, ...,  0,  0,  0],
       [ 0,  0,  0, ...,  0,  0,  0]])

In [29]:
from sklearn.metrics import accuracy_score

In [30]:
accuracy_score(y_test, y_pred)

0.54410000000000003

In [31]:
y_pred

array(['JERSEY CITY', 'CLEVELAND', 'PITTSBURGH', ..., 'JERSEY CITY',
       'BOSTON', 'BOSTON'], dtype=object)

### KNN

In [32]:
from sklearn.neighbors import KNeighborsClassifier

In [33]:
knn = KNeighborsClassifier()

In [35]:
knn.fit(x_train,y_train)

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
           metric_params=None, n_jobs=1, n_neighbors=5, p=2,
           weights='uniform')

In [36]:
y_pred1 = knn.predict(x_test)

In [37]:
accuracy_score(y_test, y_pred1)

0.32119999999999999

### Naive Bayes

In [38]:
from sklearn.naive_bayes import GaussianNB

In [39]:
nb = GaussianNB()

In [40]:
nb.fit(x_train, y_train)

GaussianNB(priors=None)

In [42]:
y_pred2 = nb.predict(x_test)

In [43]:
accuracy_score(y_test, y_pred2)

0.23563333333333333

### Deceision Tree 

In [44]:
from sklearn.tree import DecisionTreeClassifier

In [45]:
dt = DecisionTreeClassifier()

In [46]:
dt.fit(x_train, y_train)

DecisionTreeClassifier(class_weight=None, criterion='gini', max_depth=None,
            max_features=None, max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, presort=False, random_state=None,
            splitter='best')

In [47]:
y_pred3 = dt.predict(x_test)

In [48]:
accuracy_score(y_test, y_pred3)

0.48236666666666667

### Randon Forest

In [49]:
from sklearn.ensemble import RandomForestClassifier

In [50]:
rf = DecisionTreeClassifier()

In [51]:
rf.fit(x_train, y_train)

DecisionTreeClassifier(class_weight=None, criterion='gini', max_depth=None,
            max_features=None, max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, presort=False, random_state=None,
            splitter='best')

In [52]:
y_pred4 = rf.predict(x_test)

In [53]:
accuracy_score(y_test, y_pred4)

0.48296666666666666

### SVM

In [54]:
from sklearn.svm import SVC

In [56]:
sv = SVC(kernel = 'rbf', random_state = 0)

In [57]:
sv.fit(x_train, y_train)

SVC(C=1.0, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape='ovr', degree=3, gamma='auto', kernel='rbf',
  max_iter=-1, probability=False, random_state=0, shrinking=True,
  tol=0.001, verbose=False)

In [58]:
y_pred5 = sv.predict(x_test)

In [59]:
accuracy_score(y_test, y_pred5)

0.54659999999999997

In [60]:
sv1 = SVC(kernel = 'linear', random_state = 0)

In [61]:
sv1.fit(x_train, y_train)

SVC(C=1.0, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape='ovr', degree=3, gamma='auto', kernel='linear',
  max_iter=-1, probability=False, random_state=0, shrinking=True,
  tol=0.001, verbose=False)

In [62]:
y_pred6 = sv1.predict(x_test)

In [63]:
accuracy_score(y_test, y_pred6)

0.57199999999999995

### Accuracies (for 1 lakh data points)

Logistic Regression    = 0.54

KNN                    = 0.32

Naive Bayes            = 0.23

Random Forest          = 0.4829

Decision Tree          = 0.4823

SVM                    =  0.54 (rbf kernek) & 0.57 (linear kernel)